### ISSUE TITLE
[3.3.5] XT-002 Deconstructor Hardmode not activating all the way

### ISSUE THREAD

**Issue by Bullfface** [TrinityCore/TrinityCore#17141]
XT-002 Deconstructor does not gain all of his hard mode abilities if hard mode is activated.

When hard mode is activated by killing the heart, XT will cast searing light (63018) and will gain his heart breaker damage/health buff but gravity bombs will not leave behind void zones (SPELL_SUMMON_VOID_ZONE = 64203) and searing light does not summon a life spark (SPELL_SUMMON_LIFE_SPARK= 64210)

Gravity bombs should leave void zones behind when the debuff falls off and searing light should summon a life spark when the debuff falls off when hard mode is active

Go to Ulduar
Begin Xt-002 encounter
Damage Xt-002 by 25% of his health to start a heart phase
Damage the Heart to 0 before the heart phase is over to activate hard mode

3.3.5

Commit hash:	636c487704f3a019f69c821e02668b05262d638d
Parent(s):	5cdc870acc

TBD version 335.61

win 7

**Comment by joschiwald** [comment]
broken spellscripts (wrong caster of the triggered spells)

**Comment by Bullfface** [comment]
Any ideas on where to look for this>? Still working on learning it

**Comment by tkrokli** [comment]
Check the following files / folders:
- https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/Ulduar/Ulduar/boss_xt002.cpp
- https://github.com/TrinityCore/TrinityCore/tree/3.3.5/src/server/scripts/Spells (folder)

Also worth checking out: https://github.com/TrinityCore/spellwork_cs

**Comment by Bullfface** [comment]
I mean within their scripts. I am cruising through and checking everything inside to try and make sense of where the potential error could be

**Comment by Bullfface** [comment]
I found the script where the summoning of extra abilities is called in the hardmode. The below script works. Had to go into Spell_dbc to change the ticking damage of the void zone to be correct (spellID 46264)
Also found a bug where XT doesnt cast searing light until after a first heart phase
Cleaned up some emotes and such to make it more blizz like

```
--- a/src/server/scripts/Northrend/Ulduar/Ulduar/boss_xt002.cpp
+++ b/src/server/scripts/Northrend/Ulduar/Ulduar/boss_xt002.cpp
@@ -64,6 +64,7 @@ enum Spells
 
     //------------------VOID ZONE--------------------
     SPELL_VOID_ZONE                             = 64203,
+	SPELL_VOID_ZONE_PERIODIC                    = 46262,
```
```
@@ -230,7 +237,7 @@ class boss_xt002 : public CreatureScript

                events.ScheduleEvent(EVENT_ENRAGE, TIMER_ENRAGE);
                events.ScheduleEvent(EVENT_GRAVITY_BOMB, TIMER_GRAVITY_BOMB);
 -              events.ScheduleEvent(EVENT_SEARING_LIGHT, TIMER_SEARING_LIGHT);
+               events.ScheduleEvent(EVENT_SEARING_LIGHT, TIMER_SEARING_LIGHT / 2);
                //Tantrum is cast a bit slower the first time.
                events.ScheduleEvent(EVENT_TYMPANIC_TANTRUM, urand(TIMER_TYMPANIC_TANTRUM_MIN, TIMER_TYMPANIC_TANTRUM_MAX) * 2);
```

```
@@ -305,6 +316,7 @@ class boss_xt002 : public CreatureScript
                            DoCast(me, SPELL_ENRAGE);
                            break;
                        case EVENT_ENTER_HARD_MODE:
+			   Talk(EMOTE_HEART_CLOSED);
                            me->SetFullHealth();
                            DoCast(me, SPELL_HEARTBREAK, true);
                            me->AddLootMode(LOOT_MODE_HARD_MODE_1);
```
```
@@ -395,7 +405,6 @@ class boss_xt002 : public CreatureScript
            void SetPhaseOne()
            {
                Talk(SAY_HEART_CLOSED);
 -              Talk(EMOTE_HEART_CLOSED);

                me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
                me->SetReactState(REACT_AGGRESSIVE);
```
```
@@ -808,7 +817,7 @@ class spell_xt002_searing_light_spawn_life_spark : public SpellScriptLoader
                if (Player* player = GetOwner()->ToPlayer())
                    if (Unit* xt002 = GetCaster())
                        if (xt002->HasAura(aurEff->GetAmount()))   // Heartbreak aura indicating hard mode
-                           player->CastSpell(player, SPELL_SUMMON_LIFE_SPARK, true);
+                           xt002->CastSpell(player, SPELL_SUMMON_LIFE_SPARK, true);
            }

            void Register() override
@@ -844,7 +853,7 @@ class spell_xt002_gravity_bomb_aura : public SpellScriptLoader
                if (Player* player = GetOwner()->ToPlayer())
                    if (Unit* xt002 = GetCaster())
                        if (xt002->HasAura(aurEff->GetAmount()))   // Heartbreak aura indicating hard mode
-                           player->CastSpell(player, SPELL_SUMMON_VOID_ZONE, true);
+                           xt002->CastSpell(player, SPELL_SUMMON_VOID_ZONE, true);
            }

            void OnPeriodic(AuraEffect const* aurEff)
```
```
@@ -907,6 +916,37 @@ class spell_xt002_gravity_bomb_damage : public SpellScriptLoader
+        }
+};
+
+/*-------------------------------------------------------
+*
+*        VOID ZONE
+*
+*///----------------------------------------------------
+class npc_xt_void_zone : public CreatureScript
+{
+	public:
+		npc_xt_void_zone() : CreatureScript("npc_xt_void_zone") { }
+
+		struct npc_xt_void_zoneAI : public PassiveAI
+		{
+			npc_xt_void_zoneAI(Creature* creature) : PassiveAI(creature)
+			{
+				Initialize();
+			}
+			void Initialize() {
+				DoCast(me, SPELL_VOID_ZONE_PERIODIC);
+			}
+
+			void UpdateAI(uint32 /*diff*/) override { }
+		};
+
+		CreatureAI* GetAI(Creature* creature) const override
+		{
+			return new npc_xt_void_zoneAI(creature);
+		}
+
+};


class spell_xt002_heart_overload_periodic : public SpellScriptLoader
{
    public:
@@ -1088,6 +1128,7 @@ void AddSC_boss_xt002()
    new npc_boombot();

    new npc_life_spark();
+  new npc_xt_void_zone();
    new boss_xt002();

    new spell_xt002_searing_light_spawn_life_spark();

```

**Comment by Keader** [comment]
use ``` diff :)

**Comment by tkrokli** [comment]
for diff red/green highlighting: use ```diff
(no space between the 3 back ticks and the keyword)

**Comment by Bullfface** [comment]
okay that should make it look easier

**Comment by Aokromes** [comment]
plz create a pr for c++ fixes.

**Comment by Keader** [comment]
Test with #17900 ^^

**Comment by Treeston** [comment]
Closed by #17900 in 1f83412.
